raft: extract 'tracker' package

tbg opened the pull request:
First four comments are https://github.com/etcd-io/etcd/pull/10779.
Please don't review them here.

----

Create a new `tracker` package, clean up and add comments. The code movement is
in the first commit and is largely mechanical, with a priority on avoiding
anything that reviewers may want to see but wouldn't spot in the midst of the
code movement.

You'll want to hold off on nits after you've also looked at the second commit
(which addresses lots of potential nits and generally tries to leave things in
a good place).

I can squash this before merging if requested.

tbg commented:
@bdarnell, friendly ping.

bdarnell commented on raft/tracker/progress.go:
```diff
@@ -101,70 +107,86 @@ func (pr *Progress) BecomeProbe() {
 		pr.Match = n
 		updated = true
-		pr.Resume()
+		pr.RecentlyContacted = false
```
I think inlining this assignment obscures what's going on - the contact hasn't become any less recent. What's happening is that we've received an ack to our probe. I think it would be better to indirect through a method, called either Resume as before or something like ProbeReceived().

tbg replied:
I inlined this because I found `Resume()` opaque, but I like `ProbeReceived()`. Actually `ProbeAcked()` is even better I think, going with that.

bdarnell commented on raft/tracker/state.go:
```diff
@@ -14,11 +14,22 @@
+	// is necessary as the follower will react with a hint. Followers that are
+	// probed over extended periods of time are often offline.
 	StateProbe StateType = iota
+	// StateReplicate is the state steady in which a follower receives a steady
```
s/state steady/steady state/

tbg replied:
Done (refined the sentence a bit while I was there).

tbg commented on raft/tracker/progress.go:
```diff
@@ -48,7 +45,12 @@ type Progress struct {
+	// RecentlyContacted is used in StateProbe.
+	// When RecentlyContacted is true, raft should pause sending replication
+	// message to this peer until RecentlyContacted is reset. See IsPaused().
+	RecentlyContacted bool
```
Good call, I like ProbeSent because it indicates what happens and also makes it not surprising that this is reset periodically (to send another probe).

tbg commented:
Thanks for the review! I just rebased and addressed your comments in fixup commits (Planning to squash the whole thing). I'll leave this open until later tomorrow so that you (or anyone else who feels like it) can give it another look.

tbg commented:
Squashed, updated commit message, green. Merging.
